VECTOR-SEARCH-73: Increase SAI max_string_term_size_kb and max_frozen_term_size_kb to 8KiB

adelapena opened the pull request:
This PR sets 8KiB as the default for `cassandra.sai.max_string_term_size_kb`, instead of 1KiB.

It also increases the value of `cassandra.sai.max_frozen_term_size_kb` from 5KiB to the same 8KiB, because I think that the max term size for frozen collections should be at least as large as the one for individual items. I wonder if we should use an even larger value for frozen collections.

There are some additional tests, particularly for `cassandra.sai.max_vector_term_size_kb`. We aren't modifying that default here, but it seemed not covered by tests similar to the ones for the other term size limits.

By the way, `cassandra.sai.max_vector_term_size_kb` implictily sets a limit on the number of dimensions that an indexed vector column can have, if the vector subtype has fixed width. The default max term size of 16KiB sets a limit of 4096 dimensions. So maybe we could reject `CREATE INDEX` on vectors with a number of dimensions that is going to trigger `cassandra.sai.max_vector_term_size_kb` on all writes?

michaeljmarshall commented:
> So maybe we could reject `CREATE INDEX` on vectors with a number of dimensions that is going to trigger `cassandra.sai.max_vector_term_size_kb` on all writes?

Great point, I agree that it makes sense to set a dimension limit. We don't do that yet, AFAICT.

michaeljmarshall reviewed (approved):
LGTM

michaeljmarshall commented on src/java/org/apache/cassandra/index/sai/IndexContext.java:
```diff
@@ -92,8 +91,8 @@ public class IndexContext
+    public static final int MAX_STRING_TERM_SIZE = Integer.getInteger("cassandra.sai.max_string_term_size_kb", 8) * 1024;
+    public static final int MAX_FROZEN_TERM_SIZE = Integer.getInteger("cassandra.sai.max_frozen_term_size_kb", 8) * 1024;
     public static final int MAX_VECTOR_TERM_SIZE = Integer.getInteger("cassandra.sai.max_vector_term_size_kb", 16) * 1024;
     public static final int MAX_ANALYZED_SIZE = Integer.getInteger("cassandra.sai.max_analyzed_size_kb", 5) * 1024;
```
I set this a few months ago based on the `MAX_FROZEN_TERM_SIZE`. It might make sense to increase this to 8, though we can also wait until it is requested.

adelapena commented:
@michaeljmarshall thanks for reviewing.

I have updated the default value of `MAX_ANALYZED_SIZE` to 8KiB.

I have found that the check in `validateCumulativeAnalyzedTermLimit` was throwing an error if the term size was greater than on equals than `MAX_ANALYZED_SIZE`. I understand that terms with a size exactly equal to `MAX_ANALYZED_SIZE` should be accepted, as we do with other limits, so I have changed that.

> Great point, I agree that it makes sense to set a dimension limit. We don't do that yet, AFAICT.

We have an optional guardrail for the number of dimensions, `Guardrails.vectorDimensions`. It defaults to 8192. Howeverm `MAX_VECTOR_TERM_SIZE` sets an implicit limit too, which is 4096 dimensions for the default 16KiB max term size. I'm including a check on `CREATE INDEX` to also reject indexes over that limit imposed by the term size. Those two limits are independent, but it might make sense to give them consistent defaults.

I have also randomly found on the CI run above that `CollectionIndexingTest.indexEmptyMaps` is flaky because it doesn't wait for the index. I'm including that missing wait, even when it's unrelated to the changes.

adelapena commented:
@michaeljmarshall are we ready to merge this one, or will you review the last changes?

tatu-at-datastax commented:
With Data API GA (1.0.0) being released, it would be great to get this merged, for eventual deploy with Astra.
Currently some users get obscure failure for allegedly legal String values on insertion by Data API.

adelapena commented:
@michaeljmarshall I have just pushed a (trivially) rebased version of the patch. Please let me know if you are ok with [the changes](https://github.com/datastax/cassandra/pull/925#issuecomment-1880986404) done after the last review round.

michaeljmarshall reviewed (approved):
LGTM, we can merge this once tests pass

adelapena commented:
Thanks. It seems the only failure is for `testGeoAndANNOnSameColumn`, and it seems to reproduce on the base, unpatched branch.

michaeljmarshall commented:
Yes, there is a comment on the text explaining the issue. Let's merge.

adelapena commented:
Merged, thanks.
